// common_ui_components/app/pruforce/components/infoDisplay/StepChecker.tsx
import React, { FC } from 'react';
import { TouchableOpacity, Text, Image, ViewProps, View, ImageBackground } from 'react-native';
import { getImages } from '../../assets';
import { Colors, Fonts } from '../../style';

type StepCheckerPayload = {
  text?: string;
  onPress?: () => void;
};
type StepCheckerProps = ViewProps & {
  payload: StepCheckerPayload;
};

const StepChecker: FC<StepCheckerProps> = ({ payload, ...props }) => {
  const { onPress, text } = payload;

  const { style } = props;

  const upperHeight = 32;
  const upperMargin = 7;
  const iconSize = 24;
  
  return (
    <View style={{
      flexDirection: "column",
      alignSelf: "center",
    }}>
      <View style={{
        width: 57,
        height: upperHeight,
        alignSelf: "center",
      }} >
        <ImageBackground style={{
          width: 57,
          height: upperHeight,
          alignItems: "center",
        }} source={getImages.progressDialog} >
          <View style={{
            flexDirection: "row",
            height: "100%",
            width: "100%",
            alignItems: "center",
            paddingBottom: 6,
          }}>
            <Image style={{
              position: "absolute",
              width: 18,
              height: 18,
              start: 10,
              top: 5,
            }} source={getImages.progressTick} />
            <Text style={{
              fontFamily: 'Montserrat',
              fontWeight: "700",
              fontSize: 10,
              lineHeight: 12,
              position: "absolute",
              start: 34,
              top: 8,
            }}>{"1"}</Text>
          </View>
        </ImageBackground>
      </View>
      <View style={{
        alignSelf: "center",
        marginTop: upperMargin,
        marginBottom: 10,
      }} >
        <Image style={{
          width: iconSize,
          height: iconSize,
          resizeMode: "contain",
        }} source={getImages.progressFuture} />
      </View>
      <View style={{
        height: 31,
        backgroundColor: Colors.yellow,
        alignSelf: "center",
      }} >
        <Text>{"In Progress"}</Text>
      </View>

    </View>
  );
};

export default StepChecker;
