// common_ui_components/app/pruforce/components/notification/Tag.tsx
import React, { useState, useEffect } from 'react';
import { View, ViewProps } from 'react-native';
import { TagButton, TagButtonProps } from '../buttons/TagButton';

type TagPayload = {
  dataSet: Array<TagButtonProps>;
  itemHorizontalSpace? :number;
  itemVeriticalSpace? :number;
};
type TagProps = ViewProps & {
  payload: TagPayload;
};

export function useTag({ payload, ...props }: TagProps) {
  const { dataSet, itemHorizontalSpace, itemVeriticalSpace } = payload;
  const { style } = props;
  const [value, setValue] = useState<Array<TagButtonProps>>(dataSet);
  const layout = (
    <View
      style={[
        {
          flexDirection: 'row',
          flexWrap: 'wrap',
          alignItems: 'center',
        },
        style,
      ]}>
      {value.map((item) => {
        const { style } = item;
        return <TagButton payload={item.payload} style={[{
            marginRight: itemHorizontalSpace ?? 4,
            marginBottom: itemVeriticalSpace ?? 2,
        }, style]} />;
      })}
    </View>
  );
  return [layout, value, setValue];
}
