// common_ui_components/app/theme/text.ts
import { TextStyle } from 'react-native';

export const textStyleSet = {
  text_text_24_bold: {
    fontSize: 24,
    fontWeight: '700',
    lineHeight: 32,
  } as TextStyle,

  text_20_med: {
    fontSize: 20,
    fontWeight: '500',
    lineHeight: 28,
  } as TextStyle,

  text_20_reg: {
    fontSize: 20,
    fontWeight: '400',
    lineHeight: 28,
  } as TextStyle,

  text_18_bold: {
    fontSize: 18,
    fontWeight: '700',
    lineHeight: 26,
  } as TextStyle,

  text_18_med: {
    fontSize: 18,
    fontWeight: '500',
    lineHeight: 26,
  } as TextStyle,

  text_16_bold: {
    fontSize: 16,
    fontWeight: '700',
    lineHeight: 24,
  } as TextStyle,

  text_16_med: {
    fontSize: 16,
    fontWeight: '500',
    lineHeight: 24,
  } as TextStyle,

  text_14_bold: {
    fontSize: 14,
    fontWeight: '700',
    lineHeight: 22,
  } as TextStyle,

  text_14_med: {
    fontSize: 14,
    fontWeight: '500',
    lineHeight: 22,
  } as TextStyle,

  text_14_reg: {
    fontSize: 14,
    fontWeight: '400',
    lineHeight: 22,
  } as TextStyle,

  text_12_bold: {
    fontSize: 12,
    fontWeight: '700',
    lineHeight: 20,
  } as TextStyle,

  text_12_med: {
    fontSize: 12,
    fontWeight: '500',
    lineHeight: 16,
  } as TextStyle,

  text_11_med: {
    fontSize: 11,
    fontWeight: '500',
    lineHeight: 19,
  } as TextStyle,

  text_11_bold: {
    fontSize: 11,
    fontWeight: '700',
    lineHeight: 19,
  } as TextStyle,

  text_10_med: {
    fontSize: 10,
    fontWeight: '500',
    lineHeight: 18,
  } as TextStyle,

  num_40: {
    fontSize: 40,
    fontWeight: '700',
    lineHeight: 49,
  } as TextStyle,

  num_30: {
    fontSize: 30,
    fontWeight: '700',
    lineHeight: 37,
  } as TextStyle,

  num_28: {
    fontSize: 28,
    fontWeight: '600',
    lineHeight: 34,
  } as TextStyle,

  num_18: {
    fontSize: 18,
    fontWeight: '600',
    lineHeight: 22,
  } as TextStyle,

  num_14: {
    fontSize: 14,
    fontWeight: '600',
    lineHeight: 27,
  } as TextStyle,
  // ============= deprecated ====================
  content_8_med: {
    fontSize: 8,
    fontWeight: '500',
    lineHeight: 18,
  } as TextStyle,
  content_10_med: {
    fontSize: 10,
    fontWeight: '600',
  } as TextStyle,
  content_12_med: {
    fontSize: 12,
    fontWeight: '600',
  } as TextStyle,
  content_14_med: {
    fontSize: 14,
    fontWeight: '500',
    lineHeight: 22,
  } as TextStyle,
  content_14_bold: {
    fontSize: 14,
    fontWeight: '600',
    lineHeight: 22,
  } as TextStyle,
  content_16_bold: {
    fontSize: 16,
    fontWeight: '700',
    lineHeight: 24,
  } as TextStyle,
  title_24_bold: {
    fontSize: 24,
    fontWeight: 'bold',
  } as TextStyle,
  content_14_med_18: {
    fontSize: 14,
    fontWeight: '600',
    lineHeight: 18,
  } as TextStyle,
  title_18_med: {
    fontSize: 18,
    fontWeight: '600',
  } as TextStyle,
  title_20_med: {
    fontSize: 20,
    fontWeight: '600',
  } as TextStyle,
  content_16_reg: {
    fontSize: 16,
    fontWeight: '500',
  } as TextStyle,
  content_16_med: {
    fontSize: 16,
    fontWeight: '600',
  } as TextStyle,
  info_14_reg_20: {
    fontSize: 14,
    fontWeight: '500',
    lineHeight: 20,
  } as TextStyle,
  info_12_med: {
    fontSize: 12,
    fontWeight: '500',
    lineHeight: 20,
  } as TextStyle,
  info_10_bold: {
    fontSize: 10,
    fontWeight: 'bold',
  } as TextStyle,
  info_12_reg: {
    fontSize: 12,
    fontWeight: '500',
  } as TextStyle,
  title_20_bold: {
    fontSize: 20,
    fontWeight: 'bold',
  } as TextStyle,
  title_12_med: {
    fontSize: 12,
    fontWeight: '600',
  } as TextStyle,
  content_14_reg: {
    fontSize: 14,
    fontWeight: '500',
  } as TextStyle,

  info_14_med: {
    fontSize: 16,
    fontWeight: '600',
  } as TextStyle,
  content_20_med: {
    fontSize: 20,
    fontWeight: '600',
  } as TextStyle,
  title_18_bold: {
    fontSize: 18,
    fontWeight: '600',
    lineHeight: 26,
  } as TextStyle,
};
