// common_ui_components/app/components/base/header/headerWith2RightButton.tsx
import * as React from 'react';
import { View, ViewStyle, TextStyle } from 'react-native';
import { useTranslation } from 'react-i18next';
import { HeaderProps } from './header.props';
import { Button } from '../button/button';
import { Text } from '../text/text';
import { spacing, color, palette } from '../../../theme';
import { TestID } from '../../test-id';
import { IconFont } from '../../icon-font/icon-font';

// static styles
const ROOT: ViewStyle = {
  flexDirection: 'row',
  paddingHorizontal: spacing.spacing20,
  alignItems: 'center',
  paddingVertical: spacing.spacing8,
  justifyContent: 'flex-start',
};
const TITLE: TextStyle = { textAlign: 'center' };
const TITLE_MIDDLE: ViewStyle = {
  flex: 1,
  justifyContent: 'center',
  paddingLeft: spacing.spacing20,
};
const LEFT: ViewStyle = { width: 32 };
const RIGHT: ViewStyle = { width: 32 };

/**
 * Header that appears on many screens. Will hold navigation buttons and screen title.
 */
export const HeaderWith2RightButton: React.FunctionComponent<HeaderProps> = (props) => {
  const { t } = useTranslation();
  const {
    onLeftPress,
    onRightPress,
    rightIcon,
    rightIcon2,
    leftIcon,
    headerText,
    headerTx,
    style,
    titleStyle,
    iconStyle,
  } = props;
  const header = headerText || (headerTx && t(headerTx)) || '';

  return (
    <View style={{ ...ROOT, ...style }}>
      {leftIcon ? (
        <Button
          testID={TestID.LeftIcon}
          accessibilityLabel={TestID.LeftIcon}
          preset="link"
          onPress={onLeftPress}>
          <IconFont name={leftIcon} color={palette.white} size={24} />
        </Button>
      ) : (
          <View style={LEFT} />
        )}
      <View style={TITLE_MIDDLE}>
        <Text style={{ ...TITLE, ...titleStyle }} text={header} />
      </View>
      {rightIcon ? (
        <Button preset="link" onPress={onRightPress} style={{ marginRight: spacing.spacing20 }}>
          {/* TODO */}
          <IconFont name={rightIcon} color={palette.white} size={24} />
        </Button>
      ) : (
          <View style={RIGHT} />
        )}

      {rightIcon2 ? (
        <Button preset="link" onPress={onRightPress} style={{ marginRight: spacing.spacing8 }}>
          {/* TODO */}
          <IconFont name={rightIcon2} color={palette.white} size={24} />
        </Button>
      ) : (
          <View style={RIGHT} />
        )}
      <View style={{
        width: 332, height: 332, backgroundColor: palette.white, opacity: 0.16, borderRadius: 166, position: 'absolute',
        right: -160,
        top: -300,
      }}>

      </View>

    </View>
  );
};
