// common_ui_components/app/components/base/header/header.props.ts
import { ViewStyle, TextStyle, ImageStyle } from 'react-native';
import { IconTypes } from '../../icon-font/fonts/__icontype';

export interface HeaderProps {
  /**
   * Main header, e.g. POWERED BY BOWSER
   */
  headerTx?: string;

  /**
   * header non-i18n
   */
  headerText?: string;

  /**
   * Icon that should appear on the left
   */
  leftIcon?: IconTypes;

  /**
   * What happens when you press the left icon
   */
  onLeftPress?(): void;

  /**
   * Right button text
   */
  rightButton?: string;

  /**
   * What happens when you press the right icon
   */
  onRightButtonPress?(): void;

  /**
   * Icon that should appear on the right
   */
  rightIcon?: IconTypes;

  /**
   * What happens when you press the right icon
   */
  onRightPress?(): void;

  /**
   * Container style overrides.
   */
  style?: ViewStyle;

  /**
   * Title style overrides.
   */
  titleStyle?: TextStyle;

  /**
   * Icon style overrides.
   */
  iconStyle?: ImageStyle;

  /**
   * Icon that should appear on the right
   */
  rightIcon2?: IconTypes;

  /**
   * What happens when you press the right icon
   */
  onRightPress2?(): void;

  /**
   * Icon that should appear on the right
   */
  rightIcon3?: IconTypes;

  /**
   * What happens when you press the right icon
   */
  onRightPress3?(): void;

  /**
   * Color of icons and buttons
   */
  color?: string;
}
