// common_ui_components/app/components/common/modal/alert-model/AlertWithoutTitleModal.tsx
import React from 'react';
import { View, Text, TouchableOpacity, StyleSheet } from 'react-native';
import { palette, textStyleSet } from 'common_ui_components/app/theme';
import { ModalStyle } from 'common_ui_components/app/components/common/modal/Modal';
import { CommonTranslateKey } from 'common_ui_components/app/translation/__translation-keys';
import { useTranslation } from 'react-i18next';
import { pruTestID } from 'common_services_frontend';
import { Colors, Font } from '../../../../style';

export interface AlertModelProps {
  title: string;
  content: string;
  onClose: () => void;
}

const styles = StyleSheet.create({
  WRAPPER: {
    backgroundColor: palette.white,
    marginHorizontal: 12,
    borderRadius: 16,
    justifyContent: 'center',
    alignItems: 'center',
    width: '90%',
    padding: 14,
  },
});

export const AlertWithoutTitleModal: React.FC<AlertModelProps & ModalStyle> = ({
  content,
  onClose,
}) => {
  const { t } = useTranslation();
  return (
    <>
      <View style={styles.WRAPPER}>
        {content?.length > 0 && (
          <Text
            {...pruTestID('AlertContent')}
            style={{
              marginTop: 12,
              ...textStyleSet.content_20_med,
              lineHeight: 22,
              color: palette.grey33,
              paddingHorizontal: 20,
              textAlign: 'center',
            }}>
            {content}
          </Text>
        )}
        <TouchableOpacity
          accessible={false}
          onPress={() => {
            if (onClose) {
              onClose();
            }
          }}
          style={{
            width: '100%',
            backgroundColor: Colors.primaryRed,
            alignItems: 'center',
            borderRadius: 10,
            padding: 14,
            marginTop: 28,
          }}>
          <Text
            {...pruTestID('AlertConfirmButton')}
            style={{
              fontFamily: Font.primary,
              fontStyle: 'normal',
              fontWeight: '600',
              fontSize: 16,
              lineHeight: 24,
              color: Colors.white,
            }}>
            {t(CommonTranslateKey.ok)}
          </Text>
        </TouchableOpacity>
      </View>
    </>
  );
};
