// common_ui_components/app/components/filter-modal/components/Radios/style.ts
import { ViewStyle, TextStyle, Dimensions } from 'react-native';
import { PruPalette, textStyleSet, LabelStyle, spacing, PruColor, Colors } from 'common_ui_components';
// import { Colors } from 'react-native/Libraries/NewAppScreen';

export const styles = {
  lineBox: {
    flexDirection: 'row',
    marginTop: spacing.spacing11,
    alignItems: 'center',
    flexWrap: 'wrap',
    paddingRight: 20,
    marginBottom: spacing.spacing12
  },
  item: {
    borderRadius: 19, 
    borderColor: PruPalette.greyf0, 
    borderWidth: 1, 
    backgroundColor: PruPalette.white, 
    marginRight: 8, 
    justifyContent: 'center', 
    paddingHorizontal: 20, 
    marginBottom: 20,
    paddingVertical: 4
  },
  itemSelected: {
    borderRadius: 19, 
    borderColor: PruPalette.red, 
    borderWidth: 1, 
    backgroundColor: PruPalette.white, 
    marginRight: 8, 
    justifyContent: 'center', 
    paddingHorizontal: 20,
    paddingVertical: 4
  },
  itemText: {
    ...textStyleSet.info_12_reg,
    color: PruPalette.grey99
  },
  itemTextSelected: {
    color: PruPalette.red
  }
};
