// common_ui_components/app/pruforce/style/Fonts.tsx
import { Platform, StyleProp, TextStyle } from 'react-native';

const fonts = {
  montserrat: {
    font40Number: {
      fontFamily: 'Montserrat',
      fontWeight: '700',
      lineHeight: 49,
      fontSize: 40,
    } as StyleProp<TextStyle>,
    font30Number: {
      fontFamily: 'Montserrat',
      fontWeight: '700',
      lineHeight: 37,
      fontSize: 30,
    } as StyleProp<TextStyle>,
    font24Number: {
      fontFamily: 'Montserrat',
      fontWeight: '600',
      lineHeight: 30,
      fontSize: 24,
    } as StyleProp<TextStyle>,
    font24Bold: {
      fontFamily: 'Montserrat',
      fontWeight: '700',
      lineHeight: 30,
      fontSize: 24,
    } as StyleProp<TextStyle>,
    font18Number: {
      fontFamily: 'Montserrat',
      fontWeight: '700',
      lineHeight: 22,
      fontSize: 18,
    } as StyleProp<TextStyle>,
    font14Number: {
      fontFamily: 'Montserrat',
      fontWeight: '600',
      lineHeight: 17,
      fontSize: 14,
    } as StyleProp<TextStyle>,
    font14Bold: {
      fontFamily: 'Montserrat',
      fontWeight: '700',
      lineHeight: 21,
      fontSize: 14,
    } as StyleProp<TextStyle>,
  },
  label: {
    font24Bold: {
      fontWeight: '600',
      lineHeight: 32,
      fontSize: 24,
    } as StyleProp<TextStyle>,
    font20Bold: {
      fontWeight: '600',
      lineHeight: 28,
      fontSize: 20,
    } as StyleProp<TextStyle>,
    font18Bold: {
      fontWeight: '600',
      lineHeight: 26,
      fontSize: 18,
    } as StyleProp<TextStyle>,
    font16Bold: {
      fontWeight: '600',
      lineHeight: 24,
      fontSize: 16,
    } as StyleProp<TextStyle>,
    font16Regular: {
      fontWeight: '400',
      lineHeight: 24,
      fontSize: 16,
    } as StyleProp<TextStyle>,
    font14Bold: {
      fontWeight: '600',
      lineHeight: 22,
      fontSize: 14,
    } as StyleProp<TextStyle>,
    font14Medium: {
      fontWeight: '500',
      lineHeight: 21,
      fontSize: 14,
    } as StyleProp<TextStyle>,
    font14Regular: {
      fontWeight: '400',
      lineHeight: 22,
      fontSize: 14,
    } as StyleProp<TextStyle>,
    font12Medium: {
      fontWeight: '500',
      lineHeight: 16,
      fontSize: 12,
    } as StyleProp<TextStyle>,
    font11Medium: {
      fontWeight: '500',
      lineHeight: 19,
      fontSize: 11,
    } as StyleProp<TextStyle>,
    font10Medium: {
      fontWeight: '500',
      lineHeight: 15,
      fontSize: 10,
    } as StyleProp<TextStyle>,
  },
  input: {
    font24Bold: {
      ...(Platform.OS === 'ios' ? { height: 32 } : {}),
      fontWeight: '600',
      fontSize: 24,
    } as StyleProp<TextStyle>,
    font20Bold: {
      ...(Platform.OS === 'ios' ? { height: 28 } : {}),
      fontWeight: '600',
      fontSize: 24,
    } as StyleProp<TextStyle>,
    font18Bold: {
      ...(Platform.OS === 'ios' ? { height: 26 } : {}),
      fontWeight: '600',
      fontSize: 18,
    } as StyleProp<TextStyle>,
    font16Bold: {
      ...(Platform.OS === 'ios' ? { height: 24 } : {}),
      fontWeight: '600',
      fontSize: 16,
    } as StyleProp<TextStyle>,
    font16Regular: {
      ...(Platform.OS === 'ios' ? { height: 24 } : {}),
      fontWeight: '400',
      fontSize: 16,
    } as StyleProp<TextStyle>,
    font14Bold: {
      ...(Platform.OS === 'ios' ? { height: 22 } : {}),
      fontWeight: '600',
      fontSize: 14,
    } as StyleProp<TextStyle>,
    font14Medium: {
      ...(Platform.OS === 'ios' ? { height: 21 } : {}),
      fontWeight: '500',
      fontSize: 14,
    } as StyleProp<TextStyle>,
    font14Regular: {
      ...(Platform.OS === 'ios' ? { height: 22 } : {}),
      fontWeight: '400',
      fontSize: 14,
    } as StyleProp<TextStyle>,
    font12Medium: {
      ...(Platform.OS === 'ios' ? { height: 16 } : {}),
      fontWeight: '500',
      fontSize: 12,
    } as StyleProp<TextStyle>,
    font11Medium: {
      ...(Platform.OS === 'ios' ? { height: 19 } : {}),
      fontWeight: '500',
      fontSize: 11,
    } as StyleProp<TextStyle>,
    font10Medium: {
      ...(Platform.OS === 'ios' ? { height: 15 } : {}),
      fontWeight: '500',
      fontSize: 10,
    } as StyleProp<TextStyle>,
  },
};

export default fonts;
