// common_ui_components/app/components/layout/ClaimsInfo.tsx
import React, { FC } from 'react';
import { View, Text, StyleSheet, Image } from 'react-native';
import { Theme, Font } from '../../style';

type Props = React.ComponentProps<typeof View> & {
  claimDate?: string;
  claimDateSync?: string;
  claimType?: string;
  plan?: string;
  claimStatus?: string;
  cost?: string;
};

const ClaimsInfo: FC<Props> = ({
  claimDate,
  claimDateSync,
  claimType,
  plan,
  claimStatus,
  cost,
  ...props
}: Props) => (
  <View>
    <View style={[styles.container]}>
      <View style={{ flex: 4 }}>
        <Text style={styles.title}>Submit Date</Text>
        <Text style={styles.date}>{claimDate}</Text>
        <View style={{ flexDirection: 'row', alignItems: 'center' }}>
          <Image source={require('../../assets/layout/policy/sync.png')} />
          <Text style={styles.dateUpdate}>{claimDateSync}</Text>
        </View>
      </View>
      <View
        style={{
          flexDirection: 'row',
          justifyContent: 'flex-end',
        }}>
        <Text style={[styles.content, { fontSize: 14, lineHeight: 14 }]}>
          More
        </Text>
        <Image
          style={{ marginStart: Theme.margin.small }}
          source={require('../../assets/layout/policy/arrow.png')}
        />
      </View>
    </View>

    <View style={[styles.container]}>
      <View style={{ flex: 4 }}>
        <Text style={[styles.title, { marginTop: 20 }]}>Type</Text>
        <Text style={styles.content}>{claimType}</Text>

        <Text style={[styles.title, { marginTop: 20 }]}>Status</Text>
        <Text style={styles.content}>{claimStatus}</Text>
      </View>
      <View style={{ flex: 5 }}>
        <Text style={[styles.title, { marginTop: 20 }]}>Plan</Text>
        <Text style={styles.content}>{plan}</Text>

        <Text style={[styles.title, { marginTop: 20 }]}>Amount</Text>
        <Text style={styles.content}>{cost}</Text>
      </View>
    </View>
  </View>
);

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    width: '100%',
  },
  title: {
    fontSize: 14,
    fontWeight: 'normal',
    fontStyle: 'normal',
    letterSpacing: 0,
    color: '#999999',
    fontFamily: Font.primary,
  },
  date: {
    lineHeight: 28,
    fontSize: 20,
    fontWeight: '600',
    fontStyle: 'normal',
    letterSpacing: 0,
    color: '#333333',
    fontFamily: Font.primary,
  },
  dateUpdate: {
    lineHeight: 17,
    fontSize: 12,
    fontWeight: 'normal',
    fontStyle: 'normal',
    letterSpacing: 0,
    color: '#999999',
    fontFamily: Font.primary,
  },
  content: {
    fontSize: 14,
    fontWeight: '500',
    fontStyle: 'normal',
    letterSpacing: 0,
    color: '#333333',
    fontFamily: Font.primary,
  },
});

export default ClaimsInfo;
