// common_ui_components/app/components/layout/ClaimsNotitleInfo.tsx
import React, { FC } from 'react';
import { View, Text, StyleSheet, Image, ViewStyle, StyleProp } from 'react-native';
import { useTranslation } from "react-i18next";

type Props = React.ComponentProps<typeof View> & {
  claimDateSync?: string;
  paymentMethod?: string;
  plan?: string;
  value?: string;
  totalValue?: string;
  recipients?: string;
  style?: StyleProp<ViewStyle>;
};

const ClaimsInfo: FC<Props> = ({
  claimDateSync,
  paymentMethod,
  plan,
  value,
  totalValue,
  recipients,
  style,
  ...props
}) => {
  const { t } = useTranslation(["Common"]);
  return (
    <View style={style}>
      <View style={[styles.container]}>
        <View style={{ flex: 1 }}>
          <Text style={styles.title}>{t("Common:amount", "Amount")}</Text>
          <Text style={styles.date}>{value}</Text>
          <View style={{ flexDirection: 'row', alignItems: 'center' }}>
            <Image source={require('../assets/layout/policy/sync.png')} />
            <Text style={styles.dateUpdate}>{claimDateSync}</Text>
          </View>
        </View>
        <View style={{ flex: 1 }}>
          <Text style={[styles.title]}>{t("Common:total_amount", "Total Amount")}</Text>
          <Text style={styles.content}>{totalValue}</Text>
        </View>
      </View>
      <View style={[styles.container]}>
        <View style={{ flex: 1 }}>
          <Text style={[styles.title, { marginTop: 20 }]}>{t("Common:recipients", "Recipients")}</Text>
          <Text style={styles.content}>{recipients}</Text>
        </View>
        <View style={{ flex: 1 }}>
          <Text style={[styles.title, { marginTop: 20 }]}>{t("Common:payment_method", "Payment Method")}</Text>
          <Text style={styles.content}>{paymentMethod}</Text>
        </View>
      </View>
    </View>
  )
}

  const styles = StyleSheet.create({
    container: {
      flexDirection: 'row',
      width: '100%',
    },
    title: {
      fontSize: 14,
      fontWeight: 'normal',
      fontStyle: 'normal',
      letterSpacing: 0,
      color: '#999999',
      fontFamily: theme.font.primary.fontFamily,
    },
    date: {
      lineHeight: 28,
      fontSize: 20,
      fontWeight: '600',
      fontStyle: 'normal',
      letterSpacing: 0,
      color: '#333333',
      fontFamily: theme.font.primary.fontFamily,
    },
    dateUpdate: {
      lineHeight: 17,
      fontSize: 12,
      fontWeight: 'normal',
      fontStyle: 'normal',
      letterSpacing: 0,
      color: '#999999',
      fontFamily: theme.font.primary.fontFamily,
    },
    content: {
      fontSize: 14,
      fontWeight: '500',
      fontStyle: 'normal',
      letterSpacing: 0,
      color: '#333333',
      fontFamily: theme.font.primary.fontFamily,
    },
  });

  export default ClaimsInfo;
