// common_ui_components/app/components-ui/button/button.presets.ts
import { ViewStyle, TextStyle } from 'react-native';
import { color, textStyleSet } from 'common_ui_components/app/theme';

/**
 * All text will start off looking like this.
 */
const BASE_VIEW: ViewStyle = {
  paddingVertical: 12,
  paddingHorizontal: 16,
  borderRadius: 12,
  justifyContent: 'center',
  alignItems: 'center',
};

const BASE_TEXT: TextStyle = {
  ...textStyleSet.text_16_bold,
};

export enum PresetEnum {
  primary = 'primary',
  secondary = 'secondary',
  tertiary = 'tertiary',
}
export type Preset = keyof typeof PresetEnum

export type AtLeastRequired<T, U extends keyof T> = Partial<T> & Required<Pick<T, U>>;
/**
 * All the variations of button text styling within the app.
 *
 * You want to customize these to whatever you need in your app.
 */
export const buttonViewPresets: AtLeastRequired<Record<Preset, ViewStyle>, 'primary' | 'secondary'> = {
  primary: { ...BASE_VIEW, backgroundColor: color.red },
  secondary: { ...BASE_VIEW, backgroundColor: color.white, borderWidth: 2, borderColor: color.greycc },
};

export const buttonTextPresets: AtLeastRequired<Record<Preset, TextStyle>, 'primary' | 'secondary'> = {
  primary: { ...BASE_TEXT, color: color.white },
  secondary: { ...BASE_TEXT, color: color.grey33 },
};

/**
 * You want to customize these to whatever you need in your app.
 */
export const disabledButtonViewPresets: AtLeastRequired<Record<Preset, ViewStyle>, 'primary' | 'secondary'> = {
  primary: { ...BASE_VIEW, backgroundColor: color.greyc1 },
  secondary: { ...BASE_VIEW, backgroundColor: color.white, borderWidth: 2, borderColor: color.greyf0 },
};

export const disabledButtonTextPresets: AtLeastRequired<Record<Preset, TextStyle>, 'primary' | 'secondary'> = {
  primary: { ...BASE_TEXT, color: color.white },
  secondary: { ...BASE_TEXT, color: color.greyc1 },
};

/**
 * A list of preset names.
 */
export type ButtonPresetNames = keyof typeof buttonViewPresets;
