// common_ui_components/app/components-ui/button/round-base-button.tsx
import * as React from 'react';
import { TouchableHighlight, View } from 'react-native';
import { mergeAll, flatten } from 'ramda';
import { ButtonProps } from 'common_ui_components/app/components-ui/button/button.props';
import { IconTypes } from 'common_ui_components/app/components/icon-font/fonts/__icontype';
import { color } from 'common_ui_components/app/theme';
import { IconFontLayer } from 'common_ui_components/app/components-ui/icon/icon-font-with-background';
import { ButtonContainer } from 'common_ui_components/app/components-ui/button/button';

const iconColor = {
  // key => preset + disabled
  primaryfalse: color.white,
  primarytrue: color.white,
  secondaryfalse: color.grey33,
  secondarytrue: color.greycc,
};

const backgroundColor = {
  // key => preset + disabled
  primaryfalse: color.red,
  primarytrue: color.greycc,
  secondaryfalse: color.transparent,
  secondarytrue: color.transparent,
};

const borderWidth = {
  // key => preset
  primary: 0,
  secondary: 1,
};

export function RoundBaseButton(
  props: Omit<ButtonProps, 'textStyle' | 'deferTip' | 'id'> & {
    size: number;
    icon: IconTypes;
    iconSize?: number;
  },
) {
  // grab the props
  const {
    preset = 'primary',
    text,
    style: styleOverride = {},
    children,
    disabled = false,
    size,
    icon,
    iconSize,
    ...rest
  } = props;
  const viewStyle = mergeAll(flatten([styleOverride]));

  return (
    <ButtonContainer
      disabled={disabled}
      style={{
        borderRadius: '50%',
        justifyContent: 'center',
        backgroundColor: backgroundColor[preset + disabled] || color.red,
        alignItems: 'center',
      }}
      {...rest}>
      <IconFontLayer
        iconName={icon}
        iconSize={iconSize}
        iconColor={iconColor[preset + disabled] ?? color.grey33}
        backgroundColor={color.transparent}
        size={size}
        borderRadius={'50%'}
        borderWidth={borderWidth[preset]}
        borderColor={(viewStyle.borderColor || color.greyf0) as string}
      />
    </ButtonContainer>
  );
}
