// common_ui_components/app/screens/event-list-screen/styles.ts
import { ViewStyle, TextStyle, Dimensions } from 'react-native';
import { palette, spacing } from '../../theme';
import { textStyleSet } from '../../theme/text';
import { LabelStyle } from '../../components/label/label';

export const FilterStyles = {
  WRAPPER: {
    width: Dimensions.get('window').width,
    paddingBottom: 20,
    borderTopLeftRadius: 12,
    borderTopRightRadius: 12,
    backgroundColor: palette.white,
  } as ViewStyle,
  header: {
    flexDirection: 'row',
    width: '100%',
    paddingHorizontal: 14,
    paddingVertical: 18,
  } as ViewStyle,
  subBox: {
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
    paddingHorizontal: 14,
    width: '100%',
  } as ViewStyle,
  title: {
    textAlign: 'center',
    ...textStyleSet.content_14_med_18,
    color: palette.grey66,
  } as TextStyle,
  lineBox: {
    flexDirection: 'row',
    marginTop: spacing.spacing11,
    alignItems: 'center',
    width: '100%',
  } as ViewStyle,
  flex1WithMargin10: {
    flex: 1,
    marginRight: 10,
  } as ViewStyle,
  flex1: {
    flex: 1,
  } as ViewStyle,
  label: {
    flex: 1,
    height: 40,
    ...textStyleSet.content_14_med,
    borderWidth: 1,
    borderRadius: 6,
    fontColor: palette.greyba,
    justifyContent: 'center',
  } as LabelStyle,
  selected: {
    fontColor: palette.cobalt,
  } as LabelStyle,
  buttonContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    // backgroundColor: palette.orange,
    width: '100%',
  } as ViewStyle,
  buttonText: {
    fontSize: 16,
    textAlign: 'center',
  } as TextStyle,
  actionButton: {
    flex: 1,
    justifyContent: 'center',
    height: 50,
    borderRadius: 10,
    marginTop: 12,
  } as ViewStyle,
  confirmButton: {
    backgroundColor: palette.red,
    paddingHorizontal: spacing.spacing14,
  } as ViewStyle,
  resetButton: {
    borderWidth: 2,
    backgroundColor: palette.white,
    marginRight: 10,
  } as ViewStyle,
  unusedButton: {
    borderWidth: 2,
    backgroundColor: palette.white,
    marginRight: 10,
  } as ViewStyle,
};
