// common_ui_components/app/components/common/checkbox-label/checkbox-label.tsx
import * as React from 'react';
import { TextStyle, View, ViewStyle } from 'react-native';
import { color, palette, textStyleSet, typography } from '../../../theme';
import { Text } from '../../text/text';
import { IconFont } from '../../icon-font/icon-font';
import { pruTestID } from 'common_services_frontend';

const CONTAINER: ViewStyle = {
  flexDirection: 'row',
  alignItems: 'center',
};

const TEXT: TextStyle = {
  fontSize: 14,
  color: color.primary,
};

export interface CheckboxLabelProps {
  text: string;
  size?: number;
  selected?: boolean;
  style?: ViewStyle;
}

/**
 * Describe your component here
 */
export function CheckboxLabel(props: CheckboxLabelProps) {
  const { style, selected, size } = props;

  return (
    <View style={[CONTAINER, style]}>
      {selected ? (
        <IconFont name="radio_selected" size={size || 20} color={palette.red} />
      ) : (
        <IconFont name="radio" size={size || 20} color={palette.grey85} />
      )}
      <Text
        {...pruTestID('sel_text')}
        style={{
          ...textStyleSet.content_16_med,
          lineHeight: 24,
          color: palette.grey33,
          marginLeft: 8,
        }}>
        {props.text}
      </Text>
    </View>
  );
}
