// common_ui_components/app/components/common/menu-popup/menu-popup.tsx
import * as React from 'react';
import { View, ViewStyle } from 'react-native';
import { Menu } from 'react-native-paper';

const CONTAINER: ViewStyle = {
  justifyContent: 'center',
};

export interface MenuPopupProps {
  children: any[];
  anchor: React.ReactElement;
  triangCursor?: React.ReactElement;
  visible: boolean;
  onDismiss: Function;
  /**
   * An optional style override useful for padding & margin.
   */
  style?: ViewStyle;

  menuStyle?: ViewStyle;
}

/**
 * Describe your component here
 */
export function PruMenuPopup(props: MenuPopupProps) {
  const { style, menuStyle } = props;
  const [visible, setVisible] = React.useState(props.visible);
  React.useEffect(() => {
    setVisible(props.visible);
  }, [props.visible]);

  const closeMenu = () => {
    setVisible(false);
    props.onDismiss();
  };
  return (
    <>
      <Menu
        style={[CONTAINER, style]}
        visible={visible}
        onDismiss={closeMenu}
        anchor={props.anchor}>
        {props.triangCursor}
        <View style={[menuStyle]}>{props.children}</View>
      </Menu>
    </>
  );
}
