// common_ui_components/app/components/common/ActivitiesNotice.tsx
import React, { FunctionComponent as Component } from 'react';
import { StyleSheet, Text, View, TouchableOpacity, ViewStyle, Image } from 'react-native';
import { palette as PruPalette, spacing } from '../../theme';

export interface ActivitiesNoticeProps {
    content: String;
    onPress?: Function;
    style?: ViewStyle;
}

export const ActivitiesNotice: Component<ActivitiesNoticeProps> = (props) => {

    return (
        <TouchableOpacity
            activeOpacity={0.9}
            onPress={() => {
                if (props.onPress) {
                    props.onPress();
                }
            }}
            style={[styles.WRAPPER, props.style]}>
            <View
                style={{
                    width: 16,
                    height: 16,
                    borderRadius: 8,
                    borderWidth: 4,
                    borderColor: 'rgba(255,255,255,.5)',
                }} >
                <View
                    style={{
                        width: 8,
                        height: 8,
                        borderRadius: 4,
                        backgroundColor: PruPalette.white,
                    }} />
            </View>
            <Text
                style={styles.Content}
                numberOfLines={1} ellipsizeMode='tail'>
                {props.content}
            </Text>

            <Image
                style={{
                    alignSelf: 'center',
                    alignContent: 'flex-end'
                }}
                source={require('../../assets/common/white_arrow.png')}
            />

        </TouchableOpacity>
    );
}

const styles = StyleSheet.create({
    WRAPPER: {
        flex: 1,
        padding: spacing.spacing12,
        flexDirection: 'row',
        backgroundColor: '#F3343F',
        height: 40,
        borderRadius: 10,
    },
    Content: {
        alignSelf: 'center',
        fontWeight: 'bold',
        flex: 1,
        color: PruPalette.white,
        marginLeft: spacing.spacing4
    },
});
