// common_ui_components/app/components/common/AdaptiveTabBar.tsx
import React, { FC } from 'react';
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  StyleProp,
  ViewStyle,
  TextStyle,
} from 'react-native';
import {
  createMaterialTopTabNavigator,
  MaterialTopTabBarProps,
} from '@react-navigation/material-top-tabs';
import { Colors } from 'common_ui_components/app/style';
import Animated from 'react-native-reanimated';
import { palette } from '../../theme';

const AdaptiveTab = createMaterialTopTabNavigator();

type NavigatorProps = React.ComponentProps<typeof AdaptiveTab.Navigator> & {
  topContainer?: JSX.Element;
  bottomContainer?: JSX.Element;
  style?: StyleProp<ViewStyle>;
  baseStyle?: StyleProp<ViewStyle>;
  barStyle?: StyleProp<ViewStyle>;
  labelStyle?: StyleProp<TextStyle>;
  tabBarBadge?: any;
};

type AdaptiveTabBarProps = React.ComponentProps<typeof AdaptiveTab.Navigator> &
  MaterialTopTabBarProps & {
    topContainer?: JSX.Element;
    bottomContainer?: JSX.Element;
    style?: StyleProp<ViewStyle>;
    baseStyle?: StyleProp<ViewStyle>;
    barStyle?: StyleProp<ViewStyle>;
    tabBarBadge?: any;
  };

const AdaptiveTabNavigator: FC<NavigatorProps> = ({ ...props }) => {
  const labelStyle: TextStyle = {
    textAlign: 'center',
    ...(props.labelStyle as object),
  };
  return (
    <AdaptiveTab.Navigator
      tabBarOptions={{
        activeTintColor: Colors.primaryRed,
        inactiveTintColor: Colors.gray,
        style: {
          backgroundColor: Colors.white,
        },
        labelStyle: labelStyle,
        indicatorStyle: {
          borderBottomColor: Colors.primaryRed,
          borderBottomWidth: 2,
        },
      }}
      {...props}>
      {props.children}
    </AdaptiveTab.Navigator>
  );
};

var preOffSet = 0;
var barWidth = 0;

const AdaptiveTabBar: FC<AdaptiveTabBarProps> = ({
  tabBarBadge,
  baseStyle,
  barStyle,
  labelStyle,
  topContainer,
  bottomContainer,
  state,
  descriptors,
  navigation,
  position,
}) => {
  return (
    <View
      style={[styles.container, baseStyle]}
      onLayout={(event) => {
        const { width } = event.nativeEvent.layout;
        barWidth = width;
      }}>
      {topContainer}
      <View style={barStyle}>
        <View style={[{ flexDirection: 'row', alignItems: 'center' }]}>
          {state.routes.map((route, index) => {
            const { options } = descriptors[route.key];
            const label =
              options.tabBarLabel !== undefined
                ? options.tabBarLabel
                : options.title !== undefined
                ? options.title
                : route.name;
            const isFocused = state.index === index;

            const onPress = () => {
              const event = navigation.emit({
                type: 'tabPress',
                target: route.key,
                canPreventDefault: true,
              });

              if (!isFocused && !event.defaultPrevented) {
                navigation.navigate(route.name);
              }
            };

            const onLongPress = () => {
              navigation.emit({
                type: 'tabLongPress',
                target: route.key,
              });
            };

            const color = isFocused ? palette.red : palette.grey85;
            const tabbarItemStyle = isFocused
              ? {
                  borderBottomColor: palette.red,
                  borderBottomWidth: 3,
                }
              : {};

            return (
              <TouchableOpacity
                key={index.toString()}
                accessibilityRole="button"
                accessibilityLabel={options.tabBarAccessibilityLabel}
                testID={options.tabBarTestID}
                onPress={onPress}
                onLongPress={onLongPress}
                style={[styles.tabItem, tabbarItemStyle]}>
                <View style={[{ flexDirection: 'row' }]}>
                  <Text style={[{ color: color }, styles.tabLabel, labelStyle]}>{label}</Text>
                  {tabBarBadge && tabBarBadge[route.name] !== undefined && (
                    <View
                      style={[
                        {
                          borderRadius: 50,
                          paddingStart: 10,
                          paddingEnd: 10,
                          backgroundColor: color,
                          marginTop: 12,
                          marginBottom: 12,
                          marginStart: 5,
                          justifyContent: 'center',
                        },
                      ]}>
                      <Text style={[{ color: '#ffffff' }]}>{tabBarBadge[route.name]}</Text>
                    </View>
                  )}
                </View>
              </TouchableOpacity>
            );
          })}
        </View>
        {/* <Animated.View style={[styles.indicator]} /> */}
      </View>
      {bottomContainer}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    width: '100%',
    backgroundColor: Colors.white,
  },
  indicator: {
    backgroundColor: Colors.primaryRed,
    height: 3,
  },
  tabItem: {
    // alignItems: 'center',
    paddingBottom: 6,
    marginHorizontal: 12,
  },
  tabLabel: {
    fontSize: 12,
    fontWeight: '500',
    fontStyle: 'normal',
    lineHeight: 20,
  },
});

export { AdaptiveTabNavigator, AdaptiveTab, AdaptiveTabBar };
