// common_ui_components/app/components/common/Dialog.tsx
import React from "react";
import {
  Paragraph,
  Button,
  Portal,
  Dialog as _Dialog,
} from "react-native-paper";
import { StyleSheet } from "react-native";
import { Colors } from "../../style";
import { useTranslation } from "react-i18next";

type Props = React.ComponentProps<typeof _Dialog> & {
  customLayout?: React.ComponentType<any>;
  title: string;
  content?: string;
  onConfirm?: () => void;
  onCancel?: () => void;
};

const Dialog = ({ customLayout, title, content, onConfirm, onCancel, ...props }: Props) => {
  const { t } = useTranslation(["Common"]);
  return (
    <Portal>
      <_Dialog {...props} style={styles.container}>
        {customLayout ? customLayout :
          <>
            <_Dialog.Title style={styles.title}>{title}</_Dialog.Title>
            {content && (
              <_Dialog.Content>
                <Paragraph style={styles.content}>{content}</Paragraph>
              </_Dialog.Content>
            )}
            <_Dialog.Actions>
              <Button
                style={styles.btnConfirm}
                labelStyle={styles.btnTextOutline}
                mode={"outlined"}
                onPress={() => {
                  if (onConfirm) onConfirm();
                }}
              >
                {t("Common:confirm", "Confirm")}
              </Button>
              <Button
                style={styles.btnCancel}
                labelStyle={styles.btnText}
                mode={"contained"}
                theme={{ colors: { primary: Colors.primaryRed } }}
                onPress={() => {
                  if (onCancel) onCancel();
                }}
              >
                {t("Common:cancel", "Cancel")}
              </Button>
            </_Dialog.Actions>
          </>
        }
      </_Dialog>
    </Portal>
  )
};

const styles = StyleSheet.create({
  container: {
    alignItems: "center",
    padding: 5,
    borderRadius: 10,
    backgroundColor: "#ffffff",
  },
  title: {
    lineHeight: 16,
    fontSize: 16,
    fontWeight: "500",
    fontStyle: "normal",
    letterSpacing: 0,
    color: "#000000",
    marginBottom: 0,
  },
  content: {
    marginTop: 10,
    fontSize: 14,
    lineHeight: 14,
    fontWeight: "normal",
    fontStyle: "normal",
    letterSpacing: 0,
    color: "#000000",
    marginBottom: 10,
  },
  btnConfirm: {
    margin: 10,
    width: 152,
    height: 43,
    borderRadius: 24.5,
  },
  btnText: {
    textAlign: "center",
    fontSize: 15.4,
    marginTop: 12,
    color: "#ffffff",
  },
  btnTextOutline: {
    textAlign: "center",
    fontSize: 15.4,
    marginTop: 12,
    color: "#000000",
  },
  btnCancel: {
    width: 152,
    height: 43,
    borderRadius: 24.5,
    letterSpacing: 0,
    color: "#333333",
  },
});

export default Dialog;
