// common_ui_components/app/components/common/search-text-bar/search-text-bar.tsx
import * as React from 'react';
import { GestureResponderEvent, TouchableOpacity, View, ViewStyle } from 'react-native';
import { IconFont } from '../../icon-font/icon-font';
import { palette, textStyleSet } from '../../../theme';
import { Text } from '../../text/text';
import { TestID } from '../../test-id';
import { pruTestID } from 'common_services_frontend';
import { useTranslation } from 'react-i18next';
import { CommonTranslateKey } from 'common_ui_components';

export interface SearchTextBarProps {
  /**
   * An optional style override useful for padding & margin.
   */
  onFocus: (event: GestureResponderEvent) => void;
  placeholder?: string;
  style?: ViewStyle;
}

/**
 * Describe your component here
 */
export function SearchTextBar(props: SearchTextBarProps) {
  const { t } = useTranslation();
  const { onFocus, placeholder = t(CommonTranslateKey.search_for) } = props;
  return (
    <View
      style={{
        flexDirection: 'row',
        backgroundColor: palette.greyeb,
        borderRadius: 20,
        paddingLeft: 10,
        paddingVertical: 5,
        alignItems: 'center',
      }}>
      <IconFont name="search" style={{ marginLeft: 10 }} size={16} color={palette.greyba} />
      <TouchableOpacity
        accessible={false}
        activeOpacity={0.9}
        onPress={onFocus}
        style={{
          flex: 1,
          height: 26,
          justifyContent: 'center',
        }}>
        <Text
          {...pruTestID('searchBar')}
          style={{
            ...textStyleSet.content_14_med,
            color: palette.greyba,
            lineHeight: 22,
            paddingLeft: 4,
          }}>
          {placeholder}
        </Text>
      </TouchableOpacity>
    </View>
  );
}
