// common_ui_components/app/components/common/modal/more-modal/MoreModal.tsx
import React from 'react';
import { Dimensions, StyleSheet, View, FlatList } from 'react-native';
import { MoreMenuItem } from './MoreMenuItem';
import { uuidv4 } from 'common_services_frontend';
import { IconTypes } from '../../../icon-font/fonts/__icontype';
import { ModalStyle } from '../Modal';

const paddingHorizontal = 18;
const styles = StyleSheet.create({
  WRAPPER: {
    marginHorizontal: -paddingHorizontal, // 配合dialog宽度占屏100%
    width: Dimensions.get('window').width,
    paddingHorizontal: 20,
    paddingTop: 20,
  },
});
interface MoreItemModel {
  name: string;
  icon: IconTypes;
  route: string;
}
export interface MoreModalProps {
  children: MoreItemModel[];
  onClose?: Function;
}

export const MoreModal: React.FC<MoreModalProps & ModalStyle> = (props) => {
  const { children, onClose } = props;
  const _renderItem = ({ item, index }) => {
    return (
      <MoreMenuItem
        icon={item.icon}
        name={item.name}
        route={item.route}
        onClose={() => {
          if (onClose) {
            onClose();
          }
        }}
        key={index}
      />
    );
  };

  return (
    <>
      <View style={styles.WRAPPER}>
        <FlatList
          data={children}
          contentContainerStyle={{ alignSelf: 'center' }}
          renderItem={_renderItem}
          numColumns={4}
          scrollEnabled={false}
          showsVerticalScrollIndicator={false}
          showsHorizontalScrollIndicator={false}
          keyExtractor={(item) => item?.key || uuidv4}
        />
      </View>
    </>
  );
};
