// common_ui_components/app/components/common/modal/more-modal/MoreMenuItem.tsx
import React from 'react';
import { View, StyleSheet, Text, TouchableOpacity } from 'react-native';
import { EventRegister } from 'react-native-event-listeners';
import { IconFont } from '../../../icon-font/icon-font';
import { palette } from '../../../../theme';
import { IconTypes } from '../../../icon-font/fonts/__icontype';
import { pruTestID } from 'common_services_frontend';

export interface MoreMenuItemProps {
  icon: IconTypes;
  name: string;
  route: string;
  onClose: () => void;
}

const styles = StyleSheet.create({
  WRAPPER: {
    flexDirection: 'column',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
});

export const MoreMenuItem: React.FC<MoreMenuItemProps> = ({ icon, name, route, onClose }) => {
  return (
    <TouchableOpacity
      accessible={false}
      style={{ marginHorizontal: 10, marginBottom: 10, width: 80 }}
      onPress={() => {
        EventRegister.emit('moreMenuNavigateEvent', route);
        onClose();
      }}>
      <View {...pruTestID('MoreMenuItem')} style={styles.WRAPPER}> 
        <IconFont name={icon} color={palette.black} size={24} />
        <Text style={{ fontSize: 12, fontWeight: '400', color: palette.black }}>{name}</Text>
      </View>
    </TouchableOpacity>
  );
};
