// common_ui_components/app/pruforce/components/buttons/PrimaryButton.tsx
import React, { FC } from "react";
import {
  TouchableOpacity,
  Text,
  ViewProps,
} from "react-native";
import { Colors, Fonts } from "../../style";

type PrimaryButtonPayload = {
  text?: string;
  onPress?: () => void;
  disabled?: boolean;
  matchParent?: boolean;
};
type PrimaryButtonProps = ViewProps & {
  payload: PrimaryButtonPayload;
};

const PrimaryButton: FC<PrimaryButtonProps> = ({ payload, ...props }) => {
  const { onPress, text, disabled, matchParent } = payload;

  const { style } = props;
  const textColor = disabled ? Colors.grey85 : Colors.greyFF ;
  const buttonColor = disabled ? Colors.greyF0 : Colors.primary;
  const radius = 12;
  const paddingH = 20;
  const paddingV = 12;
  
  return (
    <TouchableOpacity
      style={[
        {
          borderRadius: radius,
          borderWidth: 0,
          borderColor: buttonColor,
          paddingHorizontal: paddingH,
          paddingVertical: paddingV,
          alignItems: "center",
          backgroundColor: buttonColor,
          alignSelf: matchParent ? "stretch" : "center",
        },
        style,
      ]}
      onPress={onPress}
      disabled={disabled}
    >
      <Text style={[
        Fonts.label.font16Bold,
        {
          color:  textColor,
        }
      ]}>{text}</Text>
    </TouchableOpacity>
  );
};

export default PrimaryButton;
