// common_ui_components/app/pruforce/components/buttons/SecondaryButton.tsx
import React, { FC } from 'react';
import { TouchableOpacity, Text, ViewProps } from 'react-native';
import { Colors, Fonts } from '../../style';

type SecondaryButtonPayload = {
  text?: string;
  onPress?: () => void;
  disabled?: boolean;
  matchParent?: boolean;
};
type SecondaryButtonProps = ViewProps & {
  payload: SecondaryButtonPayload;
};

const SecondaryButton: FC<SecondaryButtonProps> = ({ payload, ...props }) => {
  const { onPress, text, disabled, matchParent } = payload;

  const { style } = props;
  const textColor = disabled ? Colors.grey85 : Colors.dark33;
  const buttonColor = disabled ? Colors.greyFF : Colors.greyFF;
  const borderColor = Colors.greyF0;
  return (
    <TouchableOpacity
      style={[
        {
          borderRadius: 12,
          borderWidth: 2,
          borderColor: borderColor,
          paddingHorizontal: 18,
          paddingVertical: 10,
          alignItems: 'center',
          backgroundColor: buttonColor,
          alignSelf: matchParent ? 'stretch' : 'center',
        },
        style,
      ]}
      onPress={onPress}
      disabled={disabled}>
      <Text
        style={[
          Fonts.label.font16Bold,
          {
            color: textColor,
          },
        ]}>
        {text}
      </Text>
    </TouchableOpacity>
  );
};

export default SecondaryButton;
