// common_ui_components/app/pruforce/components/buttons/SmallButton.tsx
import React, { FC } from 'react';
import { ImageStyle, StyleProp, TextStyle } from 'react-native';
import { TouchableOpacity, Text, ViewProps, Image } from 'react-native';
import { Colors, Fonts } from '../../style';

type SmallButtonPayload = {
  text: string;
  textStyle?: StyleProp<TextStyle>;
  onPress?: () => void;
  disabled?: boolean;
  matchParent?: boolean;
  icon?: number;
  iconStyle?: StyleProp<ImageStyle>;
};
type SmallButtonProps = ViewProps & {
  payload: SmallButtonPayload;
};

const SmallButton: FC<SmallButtonProps> = ({ payload, ...props }) => {
  const { onPress, text, disabled, matchParent, icon, iconStyle, textStyle } = payload;

  const { style } = props;
  const textColor = disabled ? Colors.grey85 : Colors.dark33;
  const buttonColor = disabled ? Colors.greyF0 : Colors.greyFF;
  const borderColor = Colors.greyF0;
  return (
    <TouchableOpacity
      style={[
        {
          borderRadius: 40,
          borderWidth: 2,
          borderColor: borderColor,
          paddingHorizontal: 18,
          paddingVertical: 2,
          alignItems: 'center',
          backgroundColor: buttonColor,
          alignSelf: matchParent ? 'stretch' : 'center',
          flexDirection: "row",
        },
        style,
      ]}
      onPress={onPress}
      disabled={disabled}>
        {
          icon ? <Image style={[
            {
              width: 18,
              height: 18,
              marginRight: 4,
            },
            iconStyle,
          ]} source={icon} /> : null
        }
      <Text
        style={[
          Fonts.label.font14Bold,
          {
            color: textColor,
          },
          textStyle
        ]}>
        {text}
      </Text>
    </TouchableOpacity>
  );
};

export default SmallButton;
