// common_ui_components/app/pruforce/components/buttons/TagButton.tsx
import React, { FC } from 'react';
import { TouchableOpacity, Text, ViewProps } from 'react-native';
import { Colors, Fonts } from '../../style';

export type TagButtonPayload = {
  id: string | number;
  text?: string;
  onPress?: (id: string | number) => void;
  disabled?: boolean;
};
export type TagButtonProps = ViewProps & {
  payload: TagButtonPayload;
};

const TagButton: FC<TagButtonProps> = ({ payload, ...props }) => {
  const { text, disabled, id, onPress } = payload;

  const { style } = props;
  const textColor = Colors.cobalt;
  const borderColor = Colors.cobaltLight;
  const bkColor = Colors.greyFF;
  return (
    <TouchableOpacity
      style={[
        {
          borderRadius: 4,
          borderWidth: 2,
          borderColor: borderColor,
          paddingHorizontal: 6,
          paddingVertical: 2,
          alignItems: 'center',
          backgroundColor: bkColor,
        },
        style,
      ]}
      onPress={() => onPress ? onPress(id) : {}}
      disabled={disabled}>
      <Text
        style={[
          Fonts.label.font12Medium,
          {
            color: textColor,
          },
        ]}>
        {text}
      </Text>
    </TouchableOpacity>
  );
};

export { TagButton };
