// common_ui_components/app/pruforce/components/buttons/TextButton.tsx
import React, { FC } from 'react';
import { TouchableOpacity, Text, Image, ViewProps, StyleProp, ImageStyle, TextStyle } from 'react-native';
import { getImages } from '../../assets';
import { Colors, Fonts } from '../../style';

type TextButtonPayload = {
  text?: string;
  onPress?: () => void;
  icon?: number;
  iconColor?: string;
  iconOnLeft?: boolean;
  textStyle?: StyleProp<TextStyle>;
  iconStyle?: StyleProp<ImageStyle>;
  disabled?: boolean;
};
type TextButtonProps = ViewProps & {
  payload: TextButtonPayload;
};

const TextButton: FC<TextButtonProps> = ({ payload, ...props }) => {
  const { onPress, text, icon, iconColor, iconOnLeft, textStyle, iconStyle, disabled} = payload;

  const { style } = props;
  const iconSize = 18;
  const itemMargin = 4;
  return (
    <TouchableOpacity
      style={[
        {
          alignSelf: 'center',
          flexDirection: iconOnLeft ? 'row-reverse' : 'row',
          alignItems: 'center',
        },
        style,
      ]}
      disabled={disabled}
      onPress={onPress}>
      <Text
        style={[
          Fonts.label.font14Bold,
          {
            color: Colors.cobalt,
            marginStart: iconOnLeft ? itemMargin : 0,
          },
          textStyle ? textStyle : {}
        ]}>
        {text}
      </Text>
      {icon ? <Image
        source={icon}
        style={[{
          marginStart: iconOnLeft ? 0 : itemMargin,
          width: iconSize,
          height: iconSize,
          tintColor: iconColor ?? Colors.cobalt,
        }, iconStyle ? iconStyle : {}]}
      /> : null}
    </TouchableOpacity>
  );
};

export default TextButton;
