// common_ui_components/app/components/error-page/error-page.tsx
import React, { FunctionComponent as Component } from 'react';
import { View, ViewStyle, Text } from 'react-native';
import { useTranslation } from 'react-i18next';
import { pruTestID } from 'common_services_frontend';
import { errorPageStyles as styles } from './error-page.styles';
import { IconTypes } from '../base/icon/icons';
import { textStyleSet } from '../../theme/text';
// import { RecruitmentTranslateKey } from "recruitment_frontend";
import { TestID } from '../test-id';
import { Icon } from '../base/icon/icon';
import { CommonTranslateKey } from '../../translation/__translation-keys';
import { palette } from '../../theme';

export interface ErrorPageProps {
  icon?: IconTypes;
  textKey?: string;
  style?: ViewStyle;
  iconViewStyle?: ViewStyle;
  handleButton?: any;
  iconWidth?: number;
  iconHeight?: number;
}

export const ErrorPage: Component<ErrorPageProps> = (props) => {
  const { t } = useTranslation();
  return (
    <View style={[styles.WRAPPER, props.style]}>
      <Icon
        containerStyle={{ alignItems: 'center', ...props.iconViewStyle }}
        icon={props.icon}
        style={{ width: props.iconWidth || 100, height: props.iconHeight || 100, marginBottom: 22 }}
      />
      <Text
        {...pruTestID(TestID.IconText)}
        style={{
          ...textStyleSet.content_14_med,
          color: palette.grey85,
          paddingHorizontal: 20,
          textAlign: 'center',
        }}>
        {t(CommonTranslateKey[props.textKey || props.icon] || props.textKey) || props.textKey}
      </Text>
      <View style={{ alignItems: 'center' }}>{props.handleButton}</View>
    </View>
  );
};
