// common_ui_components/app/components/twilio-video/AttendeeName.tsx
import React, { memo } from 'react';
import { StyleSheet, View, Text } from 'react-native';
import { IconFont } from '../icon-font/icon-font';
import { pruTestID } from 'common_services_frontend';
const AttendeeName = ({
  name,
  bgColor,
  videoOn,
  audioOn,
}: {
  name: string;
  bgColor: string;
  videoOn: boolean;
  audioOn: boolean;
}) => {
  if (!name) {
    return null;
  }
  return (
    <View style={styles.rowWrap}>
      <View style={styles.rowWrap}>
        <View style={[styles.circleView, { backgroundColor: bgColor }]}>
          <Text style={styles.nameLetter}>{name[0].toUpperCase()}</Text>
        </View>
        <Text {...pruTestID('name')} numberOfLines={1} style={styles.nameText}>
          {name}
        </Text>
      </View>
      <View style={styles.rowWrap}>
        {!videoOn && <IconFont name="video_selected" color="#858585" size={24}></IconFont>}
        {!audioOn && (
          <IconFont
            name="mute_selected"
            color="#858585"
            size={24}
            style={{ marginLeft: 6 }}></IconFont>
        )}
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  rowWrap: {
    height: 54,
    flexDirection: 'row',
    alignItems: 'center',
    marginBottom: 20,
    justifyContent: 'space-between',
  },
  circleView: {
    width: 54,
    height: 54,
    borderRadius: 30,
    marginRight: 10,
  },
  nameLetter: {
    color: '#fff',
    fontSize: 24,
    textAlign: 'center',
    height: 54,
    lineHeight: 50,
  },
  nameText: {
    color: '#fff',
    fontSize: 14,
  },
});

export default memo(AttendeeName);
