// common_ui_components/app/components/twilio-video/LeaveConfirmDialog.tsx
import React from 'react';
import { Portal, Dialog } from 'react-native-paper';
import { Text } from 'react-native';
import PruButton from '../Button';
import { useTranslation } from 'react-i18next';

type Props = {
  isVisible: boolean;
  onConfirm?: () => void;
  onCancel?: () => void;
};

const LeaveConfirmDialog = ({ isVisible, onConfirm, onCancel, ...props }: Props) => {
  const { t } = useTranslation(['Common']);
  return (
    <Portal>
      <Dialog
        {...props}
        visible={isVisible}
        onDismiss={onCancel}
        style={{ padding: 10, borderRadius: 16 }}>
        <Dialog.Title style={{ textAlign: 'center', fontWeight: 'normal' }}>
          {t('Common:confirm_to_leave', 'Are you sure to leave?')}
        </Dialog.Title>
        <Dialog.Actions>
          <PruButton
            accessibilityLabel='leave'
            mode="outlined"
            style={{
              borderWidth: 1,
              backgroundColor: '#ed1b2e',
              marginBottom: 0,
            }}
            onPress={onConfirm}>
            <Text
              style={{
                color: '#fff',
                fontWeight: 'normal',
                textTransform: 'none',
              }}>
              {t('Common:leave', 'Leave')}
            </Text>
          </PruButton>
        </Dialog.Actions>

        <Dialog.Actions>
          <PruButton
            accessibilityLabel='cancel'
            mode="outlined"
            style={{ borderWidth: 1, borderColor: '#333', marginTop: 0 }}
            onPress={onCancel}>
            <Text
              style={{
                color: '#333',
                fontWeight: 'normal',
                textTransform: 'none',
              }}>
              {t('Common:cancel', 'Cancel')}
            </Text>
          </PruButton>
        </Dialog.Actions>
      </Dialog>
    </Portal>
  );
};

export default LeaveConfirmDialog;
