// common_ui_components/app/components/twilio-video/Style.ts
import { StyleSheet, Dimensions } from 'react-native';

export const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#333',
  },
  callContainer: {
    flex: 1,
    position: 'absolute',
    bottom: 0,
    top: 0,
    left: 0,
    right: 0,
  },
  button: {
    marginTop: 100,
  },

  // pop dialog start
  popupWrap: {
    marginHorizontal: -18, // dialog full screen
    width: Dimensions.get('window').width - 2,
    // paddingHorizontal: 20,
    // paddingBottom: 10,
  },
  deviceFlexWrap: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    height: '100%',
    alignItems: 'center',
  },
  deviceRow: { height: 50, paddingLeft: 20, paddingRight: 20 },

  deviceRowItem: {
    color: '#fff',
    fontSize: 14,
    marginLeft: 10,
  },
  deviceRowBorder: {
    borderBottomColor: '#666',
    borderBottomWidth: 1,
  },
  attendeeRow: {
    height: 54,
    marginBottom: 20,
  },

  // pop dialog end
  //-----------------
  identityName: {
    position: 'absolute',
    top: 82,
    left: 10,
    paddingLeft: 10,
    color: '#fff',
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
    width: 100,
    overflow: 'hidden',
  },
  remoteVideoFull: {
    flex: 1,
    width: Dimensions.get('screen').width,
    height: Dimensions.get('screen').height,
    position: 'absolute',
  },

  localVideo: {
    flex: 1,
    width: Dimensions.get('screen').width,
    height: Dimensions.get('screen').height,
    // width:200,
    // height:300,
    bottom: 0,
    right: 0,
    position: 'absolute',
    zIndex: 1,
    elevation: 1,
  },
  remoteGrid: {
    flex: 1,
    flexDirection: 'row',
    overflow: 'scroll',
    zIndex: 2,
    elevation: 2,
    marginTop: 80,
  },
  remoteVideo: {
    marginLeft: 10,
    width: 100,
    height: 100,
    // borderRadius:10
  },

  // buttom options start
  optionsContainer: {
    position: 'absolute',
    left: 0,
    bottom: 0,
    right: 0,
    height: 106,
    paddingBottom: 10,
    backgroundColor: '#222222',
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
    zIndex: 2,
    elevation: 2,
  },
  optionButton: {
    width: 52,
    height: 52,
    // marginLeft: 10,
    // marginRight: 10,
    borderRadius: 100 / 2,
    borderColor: '#666',
    borderWidth: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  leaveButton: {
    backgroundColor: '#ed1b2e',
    borderColor: '#ed1b2e',
  },
  optionColumn: { flexDirection: 'column', alignItems: 'center' },
  optionText: {
    fontSize: 12,
    color: '#fff',
    paddingTop: 4,
  },
  // buttom options end
  centeredView: {
    justifyContent: 'center',
    alignItems: 'center',
    // marginBottom:20,
    // borderWidth:1,
    height: 40,
  },
  permissionError: {
    display: 'flex',
    justifyContent: 'center',
    flexDirection: 'column',
    flex: 1,
  },
});
