// common_ui_components/app/components/recruitment/resource-item/resource-item.styles.ts
import { ViewStyle, TextStyle, ColorPropType } from 'react-native';
import { color, palette, spacing } from '../../../theme';
import { textStyleSet } from '../../../theme/text';

export const resourceItemStyles = {
  WRAPPER: {
    flex: 1,
    marginHorizontal: spacing.spacing12,
    padding: spacing.spacing12,
    flexDirection: 'row',
    backgroundColor: palette.white,
    borderRadius: 12,
  } as ViewStyle,
  Content: {
    flex: 1,
    paddingLeft: spacing.spacing12,
    flexDirection: 'column',
    justifyContent: 'flex-start',
  } as ViewStyle,
  Title: {
    paddingBottom: spacing.spacing4,
  } as ViewStyle,
  Description: {
    paddingRight: spacing.spacing14,
    paddingBottom: spacing.spacing4,
  } as ViewStyle,
  LabelContainer: {
    flexDirection: 'row',
    marginRight: spacing.spacing10,
    marginBottom: spacing.spacing4,
  } as ViewStyle,
  TitleText: {
    ...textStyleSet.content_16_med,
    lineHeight: 24,
    color: palette.grey33,
  } as TextStyle,
  DescriptionText: {
    ...textStyleSet.info_12_med,
    lineHeight: 20,
    color: palette.grey85,
  } as TextStyle,
};
