// common_ui_components/index.ts
// Styles
import { Colors, Theme, Font } from './app/style';
// translation
import { en, zhHk, vi, id } from './app/translation';
import {
  scale,
  verticalScale,
  moderateScale,
  moderateVerticalScale,
  AutoScaleStyle,
  AutoScaleStyleFitTopLeftOffset,
  guidelineBaseWidth,
  guidelineBaseHeight,
} from './app/style/ScalingUtil';

import { PruIconTypes } from './app/components';
import { assets as CommonImageAssets } from './app/assets';
import { PruUI } from './app/components-ui';

export { palette as PruPalette, color as PruColor, textStyleSet, spacing } from './app/theme';

// Components
export {
  PruDialog,
  PruListView,
  PruButton,
  PruTextInput,
  PruProvider,
  PruDefaultTheme,
  PruDarkTheme,
  PruShadow,
  PruOverlay,
  PruConfigureFonts,
  PruAvatar,
  PruList,
  PruDrawer,
  PruBadge,
  PruActivityIndicator,
  PruBanner,
  PruBottomNavigation,
  PruCard,
  PruCheckbox,
  PruChip,
  PruDataTable,
  PruDivider,
  PruFAB,
  PruHelperText,
  PruIconButton,
  PruMenu,
  PruMenuPopup,
  PruModal,
  PruPortal,
  PruProgressBar,
  PruRadioButton,
  PruSearchbar,
  PruSnackbar,
  PruSurface,
  PruSwitch,
  PruAppbar,
  PruTouchableRipple,
  PruToggleButton,
  PruCaption,
  PruHeadline,
  PruParagraph,
  PruSubheading,
  PruTitle,
  Text as PruText,
  PruListItem,
  PruListAccordion,
  PruTopTabNavigator,
  PruTab,
  PruTabBar,
  PruPopDialog,
  PruTopAdaptiveTabNavigator,
  PruAdaptiveTab,
  PruAdaptiveTabBar,
  PruListItemType,
  PruListBaseItem,
  PruListBaseItemType,
  PruExpandableIcon,
  PruTwilioVideoRoom,
  PruIconFont,
  PruIconFontLayer,
  PruErrorPage,
  PruPagingScrollView,
  PruSharePopup,
  PruSharePopupForCandidate,
  PruShareText,
  PruShareFiles,
  PruSharePDF,
  PruHeader,
  PruRedHeader,
  PruRedHeaderWithInbox,
  PruScreen,
  PruLabel,
  PruNormalButton,
  PruDeferButton,
  SearchInput as PruSearchInput,
  SearchTextBar as PruSearchTextBar,
  KeyValueItem as PruKeyValueItem,
  CustomizedKeyValueItem,
  DialogContainer as PruDialogContainer,
  Toast as PruToast,
  StatusesToast as PruStatusesToast,
  DialogToast as PruDialogToast,
  PruHeaderWith2RightButton,
  PruSearchHeader,
  PruTabView,
  ResourceList as PruResourceList,
  ScrollCategoryList,
  ResourceItem as PruResourceItem,
  MaterialPage as PruMaterialPage,
  SmartSharePopup as PruSmartSharePopup,
  TransparentHeader as PruTransparentHeader,
  HighlightText as PruHighLightText,
  ModalServiceProvider as PruModalServiceProvider,
  modalService as PruModalService,
  useModal,
  PruImageBackground,
  ExternalAvatar,
  FakeAvatar,
  LocalAvatar,
  CheckboxLabel,
  Popover,
  Triangle,
  SubmitButtonGroup,
  ActivitiesNotice,
  ChooseContactsPage as PruChooseContactsPage,
  useDropDownPicker,
  DateTimePickerView as PruDateTimePickerView,
  StaffLoginControl,
  ScollTagList as PruScrollTags,
  PruSortModal,
  PruFilterModal,
} from './app/components';

export type { PruIconTypes };
export { CommonTranslateKey } from './app/translation/__translation-keys';
export * from './app/screens';

import { commonRouterConfig } from './app/navigation/commonRouterConfig';
export {
  en,
  zhHk,
  vi,
  id,
  Colors,
  Theme,
  Font,
  scale,
  verticalScale,
  moderateScale,
  moderateVerticalScale,
  AutoScaleStyle,
  AutoScaleStyleFitTopLeftOffset,
  guidelineBaseWidth as GuidelineBaseWidth,
  guidelineBaseHeight as GuidelineBaseHeight,
  CommonImageAssets,
  commonRouterConfig,
  PruUI,
};
