// common_ui_components/app/components/common/submit-button-group/submit-button-group.tsx
import { pruTestID } from 'common_services_frontend';
import { CommonTranslateKey } from '../../../translation/__translation-keys';
import { Button as PruNormalButton } from '../../../components/base/button/button';
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { TextStyle, View, ViewStyle } from 'react-native';
import { palette, spacing, textStyleSet } from '../../../theme';
import { INACTIVE_SUBMIT_BUTTON } from 'common_ui_components/app/theme/commonStyles';
import { PruColor } from 'common_ui_components';

const CONTAINER: ViewStyle = {
  flexDirection: 'row',
  justifyContent: 'center',
  width: '100%',
};

const CONTAINER_FLAT: ViewStyle = {
  justifyContent: 'center',
  width: '100%',
  marginBottom: 20,
};

const styles = {
  buttonText: {
    fontSize: 16,
    textAlign: 'center',
  } as TextStyle,
  actionButton: {
    flex: 1,
    justifyContent: 'center',
    height: 50,
    borderRadius: 10,
    marginTop: 12,
  } as ViewStyle,
  confirmButton: {
    backgroundColor: palette.red,
    paddingHorizontal: spacing.spacing14,
  } as ViewStyle,
  cancelButton: {
    borderWidth: 2,
    borderColor: palette.greye6,
    paddingHorizontal: spacing.spacing14,
    backgroundColor: palette.white,
    marginRight: 10,
  } as ViewStyle,
  flatCancelButton: {
    borderWidth: 2,
    borderColor: palette.greyf0,
    paddingHorizontal: spacing.spacing14,
    backgroundColor: palette.white,
  } as ViewStyle,
};

export interface SubmitButtonGroupProps {
  onCancel: Function;
  onConfirm: Function;
  style?: ViewStyle;
  cancelButtonText?: string;
  confirmButtonText?: string;
}

/**
 * Describe your component here
 */
export function SubmitButtonGroup(props: SubmitButtonGroupProps) {
  const { style, onCancel, onConfirm } = props;
  const { t } = useTranslation();
  return (
    <View style={[CONTAINER, style]}>
      <PruNormalButton
        text={props.cancelButtonText || t(CommonTranslateKey.cancel)}
        onPress={() => {
          onCancel();
        }}
        textStyle={{ ...textStyleSet.text_16_bold, color: palette.grey33 }}
        style={{ ...styles.actionButton, ...styles.cancelButton }}
      />
      <PruNormalButton
        {...pruTestID('ConfirmButton')}
        text={props.confirmButtonText || t(CommonTranslateKey.confirm)}
        onPress={() => {
          onConfirm();
        }}
        textStyle={{ ...textStyleSet.text_16_bold }}
        style={{ ...styles.actionButton, ...styles.confirmButton }}
      />
    </View>
  );
}

/**
 * Describe your component here
 */
export function FlatSubmitButtonGroup(
  props: SubmitButtonGroupProps & { enableConfirmButton: boolean },
) {
  const { style, onCancel, onConfirm, enableConfirmButton } = props;
  const { t } = useTranslation();
  return (
    <View style={[CONTAINER_FLAT, style]}>
      <PruNormalButton
        {...pruTestID('ConfirmButton')}
        text={props.confirmButtonText || t(CommonTranslateKey.confirm)}
        onPress={() => {
          if (enableConfirmButton) {
            onConfirm();
          }
        }}
        textStyle={{
          ...textStyleSet.text_16_bold,
          color: enableConfirmButton ? PruColor.white : PruColor.greycf,
        }}
        style={{
          ...styles.actionButton,
          ...styles.confirmButton,
          flex: 0,
          ...(enableConfirmButton ? {} : INACTIVE_SUBMIT_BUTTON),
        }}
      />
      <PruNormalButton
        text={props.cancelButtonText || t(CommonTranslateKey.cancel)}
        onPress={() => {
          onCancel();
        }}
        textStyle={{ ...textStyleSet.text_16_bold, color: palette.grey33 }}
        style={{ ...styles.actionButton, ...styles.flatCancelButton, flex: 0 }}
      />
    </View>
  );
}
