// common_ui_components/app/pruforce/components/headers/PruHeaderContainer.tsx
import React, { FC } from 'react';
import { View, ViewProps } from 'react-native';
import LinearGradient from 'react-native-linear-gradient';

type PruHeaderContainerPayload = {
  hideGradientBackground?: boolean;
  hideBackgroundCircle?: boolean;
  gradientBackgroundHeight?: number;
  gradientBackgroundPercentX?: number;
};
type PruHeaderContainerProps = ViewProps & {
  payload: PruHeaderContainerPayload;
};
export const PruHeaderContainer: FC<PruHeaderContainerProps> = ({ payload, ...props }) => {
  const { style } = props;
  const {
    hideGradientBackground,
    hideBackgroundCircle,
    gradientBackgroundHeight,
    gradientBackgroundPercentX,
  } = payload;
  return (
    <View
      style={[
        {
          flexDirection: 'row',
          alignItems: 'center',
          minHeight: 40,
        },
        style,
      ]}>
      {hideGradientBackground ? null : (
        <>
          <LinearGradient
            accessible={false}
            colors={['rgba(255, 255, 255, 0.2)', 'rgba(255, 255, 255, 0.0)']}
            start={{ x: 0, y: 0 }}
            end={{ x: gradientBackgroundPercentX ? gradientBackgroundPercentX : 0.6, y: 0 }}
            style={{
              width: '100%',
              height: gradientBackgroundHeight ? gradientBackgroundHeight + 100 : 300,
              position: 'absolute',
              left: 0,
              top: -100,
            }}
          />
        </>
      )}
      {hideBackgroundCircle ? null : (
        <>
          <View
            style={{
              // width: 332, height: 332, backgroundColor: palette.white, opacity: 0.16, borderRadius: 166, position: 'absolute',
              width: 332,
              height: 332,
              backgroundColor: 'rgba(255, 255, 255, 0.12)',
              borderRadius: 166,
              position: 'absolute',

              right: -160,
              top: -300,
            }}
          />
        </>
      )}
      {props.children}
    </View>
  );
};
