// common_ui_components/app/components/common/modal/share-modal/ShareModal.tsx
import { ModalStyle } from '../Modal';
import { color, palette, textStyleSet } from '../../../../theme';
import React from 'react';
import { StyleSheet, View, Text, TouchableOpacity } from 'react-native';
import { Icon } from '../../../base/icon/icon';
import { useTranslation } from 'react-i18next';
import { CommonTranslateKey } from '../../../../translation/__translation-keys';
import { pruTestID } from 'common_services_frontend';

const styles = StyleSheet.create({
  WRAPPER: {
    // marginBottom: -20,
    width: '90%',
    backgroundColor: color.transparent,
  },
});
export interface ShareModalProps {
  onClose?: Function;

  onPickWhatsApp: Function;
  onPickMore: Function;
}

export const ShareModal: React.FC<ShareModalProps & ModalStyle> = (props) => {
  const { onClose, onPickWhatsApp, onPickMore } = props;
  const { t } = useTranslation();
  return (
    <>
      <View style={styles.WRAPPER}>
        <View
          style={{
            flexDirection: 'row',
            justifyContent: 'space-evenly',
            paddingBottom: 18,
            paddingTop: 20,
            borderRadius: 10,
            backgroundColor: palette.white,
            marginBottom: 8,
          }}>
          <TouchableOpacity
            accessible={false}
            activeOpacity={0.95}
            onPress={() => {
              if (onClose) {
                onClose();
              }
              onPickWhatsApp();
            }}
            style={{
              alignItems: 'center',
            }}>
            <Icon icon="whatsapp" />
            <Text {...pruTestID('WhatsappIcon')}
              style={{
                ...textStyleSet.info_12_med,
                lineHeight: 18,
                color: palette.grey33,
                paddingTop: 8,
              }}>
              {t(CommonTranslateKey.whatsapp)}
            </Text>
          </TouchableOpacity>
          <TouchableOpacity
            accessible={false}
            activeOpacity={0.95}
            onPress={() => {
              setTimeout(() => {
                onPickMore(onClose);
              }, 500);
              if (onClose) {
                onClose();
              }
            }}
            style={{
              alignItems: 'center',
            }}>
            <Icon icon="sharemore" />
            <Text {...pruTestID('ShareMoreIcon')}
              style={{
                ...textStyleSet.info_12_med,
                lineHeight: 18,
                color: palette.grey33,
                paddingTop: 8,
              }}>
              {t(CommonTranslateKey.more)}
            </Text>
          </TouchableOpacity>
        </View>

        <TouchableOpacity
          accessible={false}
          activeOpacity={0.95}
          onPress={() => {
            if (onClose) {
              onClose();
            }
          }}
          style={{
            flexDirection: 'row',
            justifyContent: 'center',
            paddingVertical: 14,
            borderRadius: 10,
            backgroundColor: palette.white,
            marginBottom: 20,
          }}>
          <Text {...pruTestID('cancel')}
            style={{
              ...textStyleSet.content_16_med,
              lineHeight: 24,
              color: palette.blue,
            }}>
            {t(CommonTranslateKey.cancel)}
          </Text>
        </TouchableOpacity>
      </View>
    </>
  );
};
