// common_ui_components/app/components/sort-modal/sort-modal.style.ts
import { ViewStyle, TextStyle, Dimensions } from 'react-native';
import { PruPalette, textStyleSet, spacing } from 'common_ui_components';

export const styles = {
  WRAPPER: {
    width: Dimensions.get('window').width,
    paddingHorizontal: 20,
    paddingBottom: 20,
    borderTopLeftRadius: 12,
    borderTopRightRadius: 12,
    backgroundColor: PruPalette.white,
  },
  header: {
    paddingVertical: 18,
  },
  headerTitle: {
    ...textStyleSet.title_18_med,
    color: PruPalette.grey33,
  },
  headerTouch: { position: 'absolute', right: 0, top: 16 },
  lineBox: {
    flexDirection: 'row',
    marginTop: spacing.spacing11,
    alignItems: 'center',
    width: Dimensions.get('window').width,
    flexWrap: 'wrap',
    paddingRight: 20,
    marginBottom: spacing.spacing12,
  },
  itemText: {
    color: PruPalette.grey85,
    textAlign: 'center',
    alignContent: 'center',
  },
  filterTitle: {
    color: PruPalette.grey99,
    ...textStyleSet.text_14_med,
    marginVertical: spacing.spacing8,
  },
};
