// common_ui_components/app/components/list/ListItem.tsx
import React, { FC } from "react";
import { View, Text, StyleSheet, Image, StyleProp, ViewStyle } from "react-native";
import { Theme, Font } from "../../style";
import { useTranslation } from "react-i18next";
import { Item } from "react-native-paper/lib/typescript/src/components/List/List";
import { Colors } from "react-native/Libraries/NewAppScreen";

enum ItemType {
  LIST_SEARCH_RESULT = 0,
  LIST_CUSTOMER_INFO = 1,
  LIST_POLICY_MANAGE = 2,
  LIST_CUSTOMER_INFO_FULL = 3,
  LIST_AGENT_AWARD = 4,
}

type Props = React.ComponentProps<typeof View> & {
  itemType?: ItemType; // 0 == ListItemSearchResult, 1 == ListItemCustomerInfo, 1 == ListItemPolicyManage
  title?: string;
  tel?: string;
  email?: string;
  address?: string;
  replyDate?: string;
  uwText?: string;
  paText?: string;
  applicationCount?: string;
  insuranceCount?: string;
  awardYear?: string;
  awardDesc?: string;
  awardRanking?: string;
  itemStyle?: StyleProp<ViewStyle>;
};

const ListItemPolicyManage: FC<Props> = ({
  title,
  tel,
  replyDate,
  uwText,
  paText,
  itemStyle,
  ...props
}) => {
  const { t } = useTranslation(["Common"]);
  return (
    <View style={[styles.container, { flexDirection: "row" }, itemStyle]} {...props}>
      <View style={{ flex: 4 }}>
        <View style={styles.innerContainer}>
          <View style={styles.redDivider} />
          <Text style={styles.title}>{title}</Text>
        </View>
        <View style={[styles.innerContainer, { marginTop: 17 }]}>
          <View style={{ flex: 2 }}>
            <Text style={styles.subTitle}>{t("Common:tel", "Mobile Nummber")}</Text>
            <View style={styles.innerContainer}>
              <Image source={require("../../assets/layout/policy/phone.png")} />
              <Text style={styles.content}>{tel}</Text>
            </View>
          </View>
          <View style={styles.divider} />
          <View style={{ flex: 1, marginLeft: Theme.padding.small }}>
            <Text style={styles.subTitle}>{t("Common:replyDate", "Reply Date")}</Text>
            <Text style={styles.content}>{replyDate}</Text>
          </View>
        </View>
      </View>
      <View style={{ flex: 1 }}>
        <View
          style={{
            flexDirection: "row",
            justifyContent: "space-between",
            marginStart: Theme.margin.small,
          }}
        >
          <Text style={styles.subTitle}>{t("Common:uw", "UW")}</Text>
          <Text style={styles.content}>{uwText}</Text>
        </View>
        <View
          style={{
            flexDirection: "row",
            justifyContent: "space-between",
            marginStart: Theme.margin.small,
          }}
        >
          <Text style={styles.subTitle}>{t("Common:pa", "PA")}</Text>
          <Text style={styles.content}>{paText}</Text>
        </View>
      </View>
    </View>
  );
}

const ListItemCustomerInfo: FC<Props> = ({ title, tel, email, itemStyle, ...props }) => {
  const { t } = useTranslation(["Common"])
  return (
    <View style={[styles.container, itemStyle]} {...props}>
      <View style={styles.innerContainer}>
        <View style={styles.redDivider} />
        <Text style={styles.title}>{title}</Text>
      </View>
      <View style={[styles.innerContainer, { marginTop: 4 }]}>
        <View style={{ flex: 1 }}>
          <View style={[styles.innerContainer, { marginStart: 13 }]}>
            <Image source={require("../../assets/layout/policy/phone.png")} />
            <Text style={styles.content}>{tel}</Text>
          </View>
          <View style={[styles.innerContainer, { marginStart: 13 }]}>
            <Image source={require("../../assets/layout/policy/email.png")} />
            <Text style={styles.content}>{email}</Text>
          </View>
        </View>
        <View style={{ justifyContent: "flex-end" }}>
          <View style={{ flexDirection: "row" }}>
            <Text style={[styles.title, { fontSize: 14, lineHeight: 14 }]}>
              {t("Common:more", "more")}
            </Text>
            <Image
              style={{ marginStart: Theme.margin.small }}
              source={require("../../assets/layout/policy/arrow.png")}
            />
          </View>
        </View>
      </View>
    </View>
  )
}

const ListItemSearchResult: FC<Props> = ({
  title,
  tel,
  applicationCount,
  insuranceCount,
  itemStyle,
  ...props
}) => {
  const { t } = useTranslation(["Common"])
  return (
    <View style={[styles.container, itemStyle]} {...props} >
      <View style={styles.innerContainer}>
        <View style={styles.redDivider} />
        <Text style={styles.title}>{title}</Text>
      </View>
      {/* <View style={[styles.innerContainer, { marginTop: 17 }]}>
        <View style={{ flexBasis: '50%', paddingRight: 15, borderRightWidth: 1, borderRightColor: '#ECEDF0'}}>
          <Text style={{ ...styles.subTitle, textAlign: 'center' }}>{t("Common:proposal", "Proposal Search")}</Text>
          <Text style={{ ...styles.content, textAlign: 'center'} }>{applicationCount}</Text>
        </View>
        <View style={{ flexBasis: '50%', paddingLeft: 15 }}>
          <Text style={{ ...styles.subTitle, textAlign: 'center' }}>{t("Common:policy", "Policy Search")}</Text>
          <Text style={{ ...styles.content, textAlign: 'center' }}>{insuranceCount}</Text>
        </View>
      </View> */}
    </View>
  )
}

const ListItemCustomerInfoFull: FC<Props> = ({ address, title, tel, email, itemStyle, ...props }) => (
  <View style={[styles.container, itemStyle]} {...props}>
    <View style={styles.innerContainer}>
      <View style={styles.redDivider} />
      <Text style={styles.title}>{title}</Text>
    </View>
    <View style={[styles.innerContainer, { marginTop: 4 }]}>
      <View style={{ flex: 1 }}>
        <View style={[styles.innerContainer, { marginStart: 13 }]}>
          <Image source={require("../../assets/layout/policy/phone.png")} />
          <Text style={styles.fullInfoContent}>{tel}</Text>
        </View>
        <View style={[styles.innerContainer, { marginStart: 13 }]}>
          <Image source={require("../../assets/layout/policy/email.png")} />
          <Text style={styles.fullInfoContent}>{email}</Text>
        </View>
        <View style={[styles.innerContainer, { marginStart: 13 }]}>
          <Image source={require("../../assets/layout/policy/address.png")} />
          <Text style={styles.fullInfoContent}>{address}</Text>
        </View>
      </View>
    </View>
  </View>
);

const ListItemAgentAward: FC<Props> = ({
  awardYear,
  awardDesc,
  awardRanking,
  itemStyle,
  ...props
}) => {
  const { t } = useTranslation(["Common"]);
  return (
    <View style={[styles.container, { flexDirection: "row" }, itemStyle]} {...props}>
      <View style={{ flex: 1, flexDirection: 'row'}}>
        <View style={styles.redDivider} />
        <View style={{ width: '70%', paddingRight: 15, borderRightWidth: 1, borderRightColor: '#ECEDF0'}}>
          <View style={{flexDirection: 'column'}}>
            <Text style={[styles.title]}>{awardDesc}</Text>
            <Text style={[styles.subTitle]}>{awardYear}</Text>
          </View>
        </View>
        <View style={{flexDirection: 'column', justifyContent: 'center', paddingLeft: 15, paddingRight: 5, flexShrink: 1}}>
          {/* <Image source={require("../../assets/layout/policy/rank.png")} /> */}
          <Text style={{fontSize: 14, fontWeight: 'bold'}}>{t("Common:rank", "Rank")}:</Text>
          <Text style={{color: '#ED1B2E', fontSize: 13, fontWeight: 'bold'}}>{awardRanking}</Text>
        </View>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    padding: Theme.padding.medium,
    borderRadius: 10,
    backgroundColor: "#ffffff",
    shadowColor: "rgba(222, 222, 222, 0.5)",
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowRadius: 16,
    shadowOpacity: 1,
  },
  innerContainer: {
    flexDirection: "row",
  },
  bottomContainer: {
    flexDirection: "row",
    width: "100%",
    padding: 17,
  },
  redDivider: {
    width: 4,
    height: 21,
    borderRadius: 2,
    backgroundColor: "#e1001f",
  },
  divider: {
    width: 2,
    height: "100%",
    borderStyle: "solid",
    borderWidth: 1,
    borderColor: "#e6e6e6",
  },
  title: {
    marginStart: Theme.margin.small,
    lineHeight: 22,
    fontSize: 16,
    fontWeight: "500",
    fontStyle: "normal",
    letterSpacing: 0,
    fontFamily: Font.primary,
    color: "#333333",
  },
  subTitle: {
    marginStart: Theme.margin.small,
    height: 22,
    lineHeight: 17,
    fontSize: 12,
    fontWeight: "500",
    fontStyle: "normal",
    letterSpacing: 0,
    fontFamily: Font.primary,
    color: "#999999",
  },
  content: {
    marginStart: Theme.margin.small,
    height: 22,
    lineHeight: 20,
    fontSize: 14,
    fontWeight: "500",
    fontStyle: "normal",
    letterSpacing: 0,
    fontFamily: Font.primary,
    color: "#333333",
  },
  fullInfoContent: {
    marginTop: 3,
    marginStart: Theme.margin.small,
    lineHeight: 18,
    fontSize: 14,
    fontWeight: "500",
    fontStyle: "normal",
    letterSpacing: 0,
    fontFamily: Font.primary,
    color: "#333333",
  }
});

export { ItemType }
export default ({ itemType, ...props }: Props) => {
  if (itemType == ItemType.LIST_SEARCH_RESULT) {
    return <ListItemSearchResult {...props} />;
  } else if (itemType == ItemType.LIST_CUSTOMER_INFO) {
    return <ListItemCustomerInfo {...props} />;
  } else if (itemType == ItemType.LIST_CUSTOMER_INFO_FULL) {
    return <ListItemCustomerInfoFull {...props} />;
  } else if (itemType == ItemType.LIST_AGENT_AWARD) {
    return <ListItemAgentAward {...props}/>
  }else {
    return <ListItemPolicyManage {...props} />;
  }
};
