// common_ui_components/app/components/list/TableItem.tsx
import React, { FC } from 'react';
import { View, Text, StyleSheet, Image, ViewStyle, StyleProp } from 'react-native';
import { Theme, Font } from '../../style';

type Props = React.ComponentProps<typeof View> & {
  claimDateSync?: string;
  paymentMethod?: string;
  plan?: string;
  value?: string;
  totalValue?: string;
  recipients?: string;
};

const ClaimsInfo: FC<Props> = ({
  claimDateSync,
  paymentMethod,
  plan,
  value,
  totalValue,
  recipients,
  ...props
}: Props) => (
  <View>
    <View style={[styles.container]}>
      <View style={{ flex: 1 }}>
        <Text style={styles.title}>Amount</Text>
        <Text style={styles.date}>{value}</Text>
        <View style={{ flexDirection: 'row', alignItems: 'center' }}>
          <Image source={require('../../assets/layout/policy/sync.png')} />
          <Text style={styles.dateUpdate}>{claimDateSync}</Text>
        </View>
      </View>
      <View style={{ flex: 1 }}>
        <Text style={[styles.title]}>Total Amount</Text>
        <Text style={styles.content}>{totalValue}</Text>
      </View>
    </View>
    <View style={[styles.container]}>
      <View style={{ flex: 1 }}>
        <Text style={[styles.title, { marginTop: 20 }]}>Recipients</Text>
        <Text style={styles.content}>{recipients}</Text>
      </View>
      <View style={{ flex: 1 }}>
        <Text style={[styles.title, { marginTop: 20 }]}>Payment Method</Text>
        <Text style={styles.content}>{paymentMethod}</Text>
      </View>
    </View>
  </View>
);

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    width: '100%',
  },
  title: {
    fontSize: 14,
    fontWeight: 'normal',
    fontStyle: 'normal',
    letterSpacing: 0,
    color: '#999999',
    fontFamily: theme.font.primary.fontFamily,
  },
  date: {
    lineHeight: 28,
    fontSize: 20,
    fontWeight: '600',
    fontStyle: 'normal',
    letterSpacing: 0,
    color: '#333333',
    fontFamily: Theme.font.primary.fontFamily,
  },
  dateUpdate: {
    lineHeight: 17,
    fontSize: 12,
    fontWeight: 'normal',
    fontStyle: 'normal',
    letterSpacing: 0,
    color: '#999999',
    fontFamily: Theme.font.primary.fontFamily,
  },
  content: {
    fontSize: 14,
    fontWeight: '500',
    fontStyle: 'normal',
    letterSpacing: 0,
    color: '#333333',
    fontFamily: Theme.font.primary.fontFamily,
  },
});

export default ClaimsInfo;
