// common_ui_components/app/components/common/popover/popover.tsx
import * as React from 'react';
import { Dimensions, Text, TextStyle, View, ViewStyle } from 'react-native';
import { color, palette, textStyleSet, typography } from '../../../theme';
import { Tooltip } from 'react-native-elements';
import { IconFont } from '../../icon-font/icon-font';

const CONTAINER: ViewStyle = {
  justifyContent: 'center',
};

export interface PopoverProps {
  /**
   * An optional style override useful for padding & margin.
   */
  anchor?: any;
  style?: ViewStyle;
  content?: string;
  width?: number;
  backgroundColor?: string
  height?: number | string
  popoverNode?: React.ReactNode;
}

/**
 * Describe your component here
 */
export function Popover(props: PopoverProps) {
  const { style, anchor, content = '', width = (Dimensions.get('window').width * 80) / 100, backgroundColor, height = 40, popoverNode } = props;

  const popoverHeight = popoverNode && height ? height : 'auto';


  const anchorComponent = anchor || (
    <IconFont name="instruction_info" color={palette.greyba} size={16} />
  );

  const renderPopoverNode =  () => {
    return (
      <View style={{ width, height: 'auto', flexDirection: 'row', alignItems: 'center', paddingHorizontal: 15, flex: 1}}>
        {
          popoverNode
        }
      </View>
    )
  }

  return (
    <View style={[CONTAINER, style]}>
      <Tooltip
        backgroundColor={backgroundColor || palette.cobalt}
        overlayColor={color.transparent}
        width={width}
        height={height}
        containerStyle={{
          justifyContent: 'center',
          height: popoverHeight
        }}
        popover={
          popoverNode 
            ? 
              renderPopoverNode() 
            : <Text
                style={{
                  ...textStyleSet.content_14_reg,
                  lineHeight: 22,
                  color: palette.white,
                  
                }}
              >
              {content}
            </Text>
        }
      >
        {anchorComponent}
      </Tooltip>
    </View>
  );
}
