// common_ui_components/app/components/common/pop-dialog/pop-dialog.styles.ts
import { palette, spacing } from '../../../theme';
import { ViewStyle, TextStyle } from 'react-native';

export const popDialogStyles = {
  WRAPPER: {
    width: '100%',
    paddingHorizontal: 14,
  } as ViewStyle,
  title: {
    textAlign: 'center',
    marginTop: spacing.spacing30,
  } as TextStyle,
  messageContainer: {
    paddingHorizontal: spacing.spacing20,
  } as ViewStyle,
  message: {
    textAlign: 'left',
    marginTop: 9,
    marginBottom: 16,
  } as TextStyle,
  buttonText: {
    fontSize: 16,
    textAlign: 'center',
  } as TextStyle,
  confirmButton: {
    justifyContent: 'center',
    backgroundColor: palette.red,
    height: 50,
    borderRadius: 10,
    marginTop: 12,
    paddingHorizontal: spacing.spacing14,
  } as ViewStyle,
  cancelButton: {
    justifyContent: 'center',
    borderWidth: 2,
    backgroundColor: palette.white,
    height: 50,
    borderRadius: 10,
    marginTop: 12,
    paddingHorizontal: spacing.spacing14,
    marginBottom: spacing.spacing18,
  } as ViewStyle,
};
